src/server/list_family.cc

```diff
@@ -867,6 +878,47 @@ void BRPopLPush(CmdArgList args, ConnectionContext* cntx) {
   }
 }
 
+void BLMove(CmdArgList args, ConnectionContext* cntx) {
+  string_view src = ArgS(args, 1);
+  string_view dest = ArgS(args, 2);
+  string_view timeout_str = ArgS(args, 5);
+
+  float timeout;
+  if (!absl::SimpleAtof(timeout_str, &timeout)) {
+    return (*cntx)->SendError("timeout is not a float or out of range");
+  }
+
+  if (timeout < 0) {
+    return (*cntx)->SendError("timeout is negative");
+  }
+
+  ToUpper(&args[3]);
+  ToUpper(&args[4]);
+
+  optional<ListDir> src_dir = ParseDir(ArgS(args, 3));
+  optional<ListDir> dest_dir = ParseDir(ArgS(args, 4));
+  if (!src_dir || !dest_dir) {
+    return (*cntx)->SendError(kSyntaxErr);
+  }
+
+  BPopPusher bpop_pusher(src, dest, *src_dir, *dest_dir);
+  OpResult<string> op_res = bpop_pusher.Run(cntx->transaction, unsigned(timeout * 1000));
```

Comment:
The params passed to RecordJournal function in the RunSingle function should be updated. I used RPOPLPUSH, it now should be LMOVE

Reply:
I looked now at `BPopper::Run`.  I think that this code 
```
auto cb = [this](Transaction* t, EngineShard* shard) {
    Pop(t, shard);
    OpArgs op_args = t->GetOpArgs(shard);
    if (op_args.shard->journal()) {
      string command = dir_ == ListDir::LEFT ? "LPOP" : "RPOP";
      RecordJournal(op_args, command, ArgSlice{key_}, 1);
    }
    return OpStatus::OK;
  };
``` 
is incorrect @adiholden  - because we run the callback on both shards but Pop inside checks if it is a source shard.

Reply:
fixed RPOPLPUSH

Reply:
The BPopper is used only for single key command, so the pop should run only in the key shard thread no?

Reply:
no, https://redis.io/commands/blpop/   blpop eventually pops from a single key but it monitors multiple keys. Hence, it will run on multiple shards.